torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <torch/nn/module.h>
+#include <torch/nn/modules/any.h>
+
+#include <ATen/Error.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace torch { namespace nn {
+
+/// A `Sequential` module is a container for any number of other modules. Its
+/// `forward()` method chains outputs to inputs and returns the final output.
+/// The `Sequential` class reference semantics.
+class Sequential : public CloneableModule<Sequential> {
+ public:
+  using Iterator = std::vector<std::shared_ptr<AnyModule>>::iterator;
+
+  /// Constructs the `Sequential` from a pack of modules. Each module can either
+  /// be a plain value (e.g. `Linear`) or a boxed value (e.g.
+  /// `shared_ptr<Linear>`). Unboxed modules will be moved into `shared_ptr`s
+  /// internally.
+  template <
+      typename... Modules,
+      typename = disable_for_t<Sequential, Modules...>>
+  explicit Sequential(Modules&&... modules) {
+    push_back(std::forward<Modules>(modules)...);
+  }
+
+  /// Feeds the `inputs` to the first module, then chains the output of each
+  /// module with the input of the next, in order of construction.
+  template <typename ReturnType, typename... ArgumentTypes>
+  ReturnType forward(ArgumentTypes&&... arguments) {
+    AT_CHECK(!is_empty(), "Cannot call forward() on an empty Sequential");
```

Comment:
Why not? If it semantically makes sense (which I think it does), it might make sense to just return the input directly instead of erroring.

Reply:
What if there are multiple inputs?

Reply:
Do I get it right that you're special casing the first op in the sequence to allow multiple inputs, whereas further ops can only take one input? If yes, then multiple inputs and no components is probably an error case.

If there's no special case, say intermediate components can have multiple outputs, then multiple inputs and no component would be the same as having the last component output multiple outputs.

I'm not 100% sure what the best behavior is, I'm trying to make it follow semantics and avoid edge cases that need to error out.